python/lance/data/convert/nuscenes.py

```diff
@@ -0,0 +1,267 @@
+        return objects
+
+    def read_metadata(self, num_rows: int = 0):
+        """
+        Converts the dataset to a dataframe that is flattened.
+        To do so, we need to get all the samples and
+        join various entities to the sample.
+        """
+        instances_df = []
+        current_row = 0
+           
+        for sample in self.instance_data["sample"]:
+            if num_rows > current_row or num_rows == 0:
+                current_row += 1
+
+                # Find all related entities to the sample
+                log_data = self._find_foreign_object(sample["log_token"], "token", "log")
+                sample_data = self._find_foreign_object(sample["token"], "sample_token", "sample_data")
+                ego_pose = self._find_foreign_object(sample_data["ego_pose_token"], "token", "ego_pose")
+                calibrated_sensor = self._find_foreign_object(sample_data["calibrated_sensor_token"], "token", "calibrated_sensor")
+                sensor = self._find_foreign_object(calibrated_sensor["sensor_token"], "token", "sensor")
+
+                # Create the flatten sample
+                sample_joined = copy.deepcopy(sample)
+                sample_joined = self._merge_dict(sample_joined, log_data, "log")
+                sample_joined = self._merge_dict(sample_joined, sample_data, "sample_data")
+                sample_joined = self._merge_dict(sample_joined, ego_pose, "ego_pose")
+                sample_joined = self._merge_dict(sample_joined, calibrated_sensor, "calibrated_sensor")
+                sample_joined = self._merge_dict(sample_joined, sensor, "sensor")
+
+                # Find the annotations for the sample
+                object_anns = self._find_annotations(sample_data["token"], "sample_data_token", "object_ann")
+                surface_ann = self._find_annotations(sample_data["token"] , "sample_data_token", "surface_ann")
+                sample_joined["object_ann"] = object_anns
+                sample_joined["surface_ann"] = surface_ann
+                
+                instances_df.append(sample_joined)
+        
+        return pd.DataFrame(instances_df)
+    
+    def get_schema(self) -> pa.Schema:
+        """
+        Returns the Arrow schema.
+        This is flatten/denormalized for better ergonomics for
+        ML - i.e. filtering/slicing the dataset and reducing joins.
+        Column names are renamed accordingly to prevent collisions.
+        """
+        # Mask
+        mask_schema = pa.struct([
+            pa.field("size", pa.list_(pa.int32(), 2)),
+            pa.field("counts", pa.string())
+        ])
+
+        # Surface Ann
+        surface_ann_schema = pa.struct([
+            pa.field("token", pa.string()),
+            pa.field("sample_data_token", pa.string()),
+            pa.field("category_token", pa.string()),
+            pa.field("mask", mask_schema), # TODO: change to RLE once type is created 
+
+            # Category
```

Comment:
what is difference between `category_token` and `category_token_` (with a suffix)

Reply:
removed foreign/primary key tokens, except for the root table